loopchain/consensus/syncer.py

```diff
@@ -0,0 +1,160 @@
+            channel = GRPCHelper().create_client_channel(target)
+            self._stub_list.append(loopchain_pb2_grpc.PeerServiceStub(channel))
+
+    async def sync_start(self):
+        while True:
+            await self._request_block()
+            await self._raise_event()
+
+    async def _request_block(self):
+        def _request(index: int, height: int):
+            peer_stub = self._stub_list[index]
+
+            peer_stub.BlockRequest(loopchain_pb2.PeerHeight(
+                peer=ChannelProperty().peer_target,
+                channel=self._block_manager.channel_name,
+                height=height
+            ), conf.GRPC_TIMEOUT)
+
+        # Sync mode check
+        # LFT is two step process. So It need to wait for a height of at least 2.
+        gap = self._max_height_in_nodes-self.__blockchain.block_height
+        if gap < RAISE_EVENT_INTERVAL:
+            return
+
+        goal = min(conf.CITIZEN_ASYNC_RESULT_MAX_SIZE+1, self._max_height_in_nodes-RAISE_EVENT_INTERVAL)
+        for i in range(1, goal):
+            height = self.__blockchain.block_height + i
+            try:
+                if height in self._request_history_list.keys():
+                    during_request_time = time.time()-self._request_history_list[height][1]
+                    if during_request_time > conf.LFT_SYNC_REQUEST_WAIT:
+                        retry_target_index = (self._request_history_list[height][0] + 1) % len(self._stub_list)
+                        _request(retry_target_index, height)
+                        self._request_history_list[height] = [retry_target_index, time.time()]
+                else:
+                    _request(self._target_index, height)
+                    self._request_history_list[height] = [self._target_index, time.time()]
+                    self._target_index = (self._target_index+1) % len(self._stub_list)
+
+            except IndexError:
+                self._target_index = 0
+
+            await asyncio.sleep(0)
+
+    def receive_vote(self, vote: 'BlockVote'):
+        height = vote.block_height
+        self._max_height_in_nodes = max(height, self._max_height_in_nodes)
+
+        if abs(self.__blockchain.block_height-height) < conf.CITIZEN_ASYNC_RESULT_MAX_SIZE:
+            if height not in self._vote_info_other_nodes.keys():
+                self._vote_info_other_nodes[height] = []
+
+            self._vote_info_other_nodes[height].append(vote)
+
+    def receive_data(self, block_data: 'Block'):
+        height = block_data.header.height
+        self._max_height_in_nodes = max(height, self._max_height_in_nodes)
+
+        diff_height_info = abs(self.__blockchain.block_height-self._max_height_in_nodes)
+        if diff_height_info < conf.CITIZEN_ASYNC_RESULT_MAX_SIZE:
+            if height not in self._data_info_other_nodes.keys():
+                self._data_info_other_nodes[height] = []
+
+            self._data_info_other_nodes[height].append(block_data)
```

Comment:
How about extracting this lines with `append_block_if_not_exists`-some?
Because there're some lines duplicated (+ readability).

Reply:
I add `_append_block_if_not_exists` function.
It use in `receive_vote` function and `receive_data`function.